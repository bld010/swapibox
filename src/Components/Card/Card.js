import React from 'react';
import PropTypes from 'prop-types';
import './Card.scss'

const Card = (props) => {

  let { name, 
    species, 
    homeworld, 
    population, 
    language, 
    terrain, 
    model, 
    vehicle_class, 
    residents, 
    climate, 
    type,
    passengers,
    handleFavorite,
    favorites
   } = props



   let getFavoriteInfo = () => {

    if (type === 'people') {
      return { name, species, homeworld, population, language, type}
    }

    if (type === 'vehicles') {
      return { name, model, vehicle_class, passengers, type}
    }

    if (type === 'planets') {
      return { name, population, terrain, climate, residents, type}
    }

  }

  let peopleCard = 
    <>
    <h5>| Species: </h5> 
    <p>{species}</p>
    <h5>| Home: </h5>
    <p>{homeworld} (Pop. {population})</p>
    <h5>| Language: </h5>
    <p>{language}</p>
    </>;

  let residentsList = 
    <>
      <p className='residents'></p>
      <span>{residents && residents.length > 0 && residents.join(', ')}</span>
      {residents && residents.length === 0 && <p>None</p>}
    </>
  
  let planetCard = 
    <>
    <h5>| Terrain: </h5>
    <p>{terrain}</p>
    <h5>| Population: </h5>
    <p>{population}</p>
    <h5>| Climate: </h5>
    <p>{climate}</p>
    <h5>| Residents: </h5>
    {residentsList}
    </>;
  
  let vehicleCard = 
    <>
    <h5>| Model: </h5>
    <p>{model} </p>
    <h5>| Class: </h5>
    <p>{vehicle_class}</p>
    <h5>| Passengers: </h5>
    <p>{passengers}</p>
    </>;

  
  let getCardClass = () => {
    if (favorites.map(favorite => favorite.name).includes(name)) {
      return 'favorite'
    } else {
      return ''
    }
  }
  

  let handleFavoriteButton = (favoritedCard) => {
    handleFavorite(favoritedCard);
  }

  return(
    <article className={type + ' Card'}>
      <button className={getCardClass()} onClick={() => handleFavoriteButton(getFavoriteInfo())}></button>
        
      <h4>{name}</h4>
      
      {type === 'people' && peopleCard}
      {type === 'planets' && planetCard}
      {type === 'vehicles' && vehicleCard}
      
    </article>
  )
}

export default Card;

Card.propTypes = {
  type: PropTypes.string,
  name: PropTypes.string,
  species:  PropTypes.string,
  homeworld:  PropTypes.string,
  population:  PropTypes.string,
  language:  PropTypes.string,
  terrain:  PropTypes.string,
  model:  PropTypes.string,
  vehicle_class:  PropTypes.string,
  residents: PropTypes.array,
  climate:  PropTypes.string,
  passengers:  PropTypes.string,
  handleFavorite: PropTypes.func,
  favorites: PropTypes.array
}