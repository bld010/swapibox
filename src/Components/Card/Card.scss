$main-color: green;
$name-color: goldenrod;
$box-shadow: 0px 0px 8px 3px rgb(67, 165, 71);

.Card {
  height: 250px;
  width: 250px;
  border: 1px solid $main-color;
  color: $main-color;
  background: black;
  margin: 10px 30px 30px 10px;
  box-shadow: $box-shadow;
  padding: 20px;
  border-radius: 0 29px 0 29px;
  text-align: left;

  & h4 {
    color: $name-color;
    font-size: 22px;
    text-transform: uppercase;
    
  } 

  & p {
    font-size: 18px;
    text-transform: uppercase;
    font-weight: bold;
  }

  & button {
    margin-left: 220px;
    border-radius: 50%;
    border: 1px solid $main-color;
    height: 30px;
    width: 30px;
    background: black;
    cursor: pointer
  }
}