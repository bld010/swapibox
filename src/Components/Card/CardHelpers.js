export let getFavoriteInfo = (props) => {

  let { name, 
    species, 
    homeworld, 
    population, 
    language, 
    terrain, 
    model, 
    vehicle_class, 
    residents, 
    climate, 
    type,
    passengers,
   } = props

  if (type === 'people') {
    return { name, species, homeworld, population, language, type}
  }

  if (type === 'vehicles') {
    return { name, model, vehicle_class, passengers, type}
  }

  if (type === 'planets') {
    return { name, population, terrain, climate, residents, type}
  }

}

export let getCardClass = (favorites, name) => {
  if (favorites.map(favorite => favorite.name).includes(name)) {
    return 'favorite'
  } else {
    return ''
  }
}