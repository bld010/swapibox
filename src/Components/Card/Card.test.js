import React from 'react';
import Card from './Card';
import { shallow } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

let favorites = [{name: 'Han Solo'}, {name: 'X-Wing'}, {name: 'Hoth'}]

describe('Card', () => {
  describe('snapshots', () => {

      it('should match the people snapshot if a favorited person card', () => {
        const wrapper = shallow(<Card 
            name="Han Solo"
            species="Human"
            homeworld="Tatooine"
            population="3"
            language="Galactic Basic"
            type="people"
            favorites={favorites} />
          )
          expect(wrapper).toMatchSnapshot();
        })

          it('should match the people snapshot if not favorited person card', () => {
            const wrapper = shallow(<Card 
                name="Anakin Skywalker"
                species="Human"
                homeworld="Tatooine"
                population="3"
                language="Galactic Basic"
                type="people"
                favorites={favorites} />
              )
              expect(wrapper).toMatchSnapshot();    
      })

      it('should match the vehicles snapshot if not favorited vehicle card', () => {
        const wrapper = shallow(<Card 
            name="AT-AT"
            model="Giant Walking Thingy"
            vehicle_class="Real Big Vehicle"
            passengers="4"
            type="vehicles"
            favorites={favorites} />
          )
          expect(wrapper).toMatchSnapshot();    
    })

      it('should match the vehicles snapshot if favorited vehicle card', () => {
        const wrapper = shallow(<Card 
            name="X-Wing"
            model="Giant Flying Thingy"
            vehicle_class="Small Jet"
            passengers="1"
            type="vehicles"
            favorites={favorites} />
          )
          expect(wrapper).toMatchSnapshot();    
    })


    it('should match the planets snapshot if favorited planet card', () => {
      const wrapper = shallow(<Card 
          name="Hoth"
          terrain="Scary"
          population="?"
          climate="Cold, probably"
          residents={[]}
          type="planets"
          favorites={favorites} />
        )
        expect(wrapper).toMatchSnapshot();    
    })


    it('should match the planets snapshot if not favorited planet card', () => {
      const wrapper = shallow(<Card 
          name="Naboo"
          terrain="Scary"
          population="?"
          type="planets"
          climate="Cold, probably"
          residents={['Joe', 'Bob', 'Sarah']}
          favorites={favorites} />
        )
        expect(wrapper).toMatchSnapshot();    
    })
  })

it('should fire handleFavorite when favorited', () => {
  const mockHandleFavorite = jest.fn();
  
  const wrapper = shallow(<Card 
    name="Naboo"
    terrain="Scary"
    population="?"
    type="planets"
    climate="Cold, probably"
    residents={['Joe', 'Bob', 'Sarah']}
    handleFavorite ={mockHandleFavorite}
    favorites={favorites} />
  )

  wrapper.find('button').simulate('click')
  expect(mockHandleFavorite).toHaveBeenCalled();
})



})