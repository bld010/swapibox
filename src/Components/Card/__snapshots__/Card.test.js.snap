// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card snapshots should match the people snapshot if a favorited person card 1`] = `
<article
  className="people Card"
>
  <button
    className="favorite"
    onClick={[Function]}
  />
  <h4>
    Han Solo
  </h4>
  <h5>
    | Species: 
  </h5>
  <p>
    Human
  </p>
  <h5>
    | Home: 
  </h5>
  <p>
    Tatooine
     (Pop. 
    3
    )
  </p>
  <h5>
    | Language: 
  </h5>
  <p>
    Galactic Basic
  </p>
</article>
`;

exports[`Card snapshots should match the people snapshot if not favorited person card 1`] = `
<article
  className="people Card"
>
  <button
    className=""
    onClick={[Function]}
  />
  <h4>
    Anakin Skywalker
  </h4>
  <h5>
    | Species: 
  </h5>
  <p>
    Human
  </p>
  <h5>
    | Home: 
  </h5>
  <p>
    Tatooine
     (Pop. 
    3
    )
  </p>
  <h5>
    | Language: 
  </h5>
  <p>
    Galactic Basic
  </p>
</article>
`;

exports[`Card snapshots should match the planets snapshot if favorited planet card 1`] = `
<article
  className="planets Card"
>
  <button
    className="favorite"
    onClick={[Function]}
  />
  <h4>
    Hoth
  </h4>
  <h5>
    | Terrain: 
  </h5>
  <p>
    Scary
  </p>
  <h5>
    | Population: 
  </h5>
  <p>
    ?
  </p>
  <h5>
    | Climate: 
  </h5>
  <p>
    Cold, probably
  </p>
  <h5>
    | Residents: 
  </h5>
  <p
    className="residents"
  />
  <span />
  <p>
    None
  </p>
</article>
`;

exports[`Card snapshots should match the planets snapshot if not favorited planet card 1`] = `
<article
  className="planets Card"
>
  <button
    className=""
    onClick={[Function]}
  />
  <h4>
    Naboo
  </h4>
  <h5>
    | Terrain: 
  </h5>
  <p>
    Scary
  </p>
  <h5>
    | Population: 
  </h5>
  <p>
    ?
  </p>
  <h5>
    | Climate: 
  </h5>
  <p>
    Cold, probably
  </p>
  <h5>
    | Residents: 
  </h5>
  <p
    className="residents"
  />
  <span>
    Joe, Bob, Sarah
  </span>
</article>
`;

exports[`Card snapshots should match the vehicles snapshot if favorited vehicle card 1`] = `
<article
  className="vehicles Card"
>
  <button
    className="favorite"
    onClick={[Function]}
  />
  <h4>
    X-Wing
  </h4>
  <h5>
    | Model: 
  </h5>
  <p>
    Giant Flying Thingy
     
  </p>
  <h5>
    | Class: 
  </h5>
  <p>
    Small Jet
  </p>
  <h5>
    | Passengers: 
  </h5>
  <p>
    1
  </p>
</article>
`;

exports[`Card snapshots should match the vehicles snapshot if not favorited vehicle card 1`] = `
<article
  className="vehicles Card"
>
  <button
    className=""
    onClick={[Function]}
  />
  <h4>
    AT-AT
  </h4>
  <h5>
    | Model: 
  </h5>
  <p>
    Giant Walking Thingy
     
  </p>
  <h5>
    | Class: 
  </h5>
  <p>
    Real Big Vehicle
  </p>
  <h5>
    | Passengers: 
  </h5>
  <p>
    4
  </p>
</article>
`;
